tests/integration_tests/conftest.py

```diff
@@ -0,0 +1,91 @@
+# This file is part of cloud-init. See LICENSE file for license information.
+import os
+import pytest
+from contextlib import contextmanager
+
+from tests.integration_tests import integration_settings
+from tests.integration_tests.platforms import (
+    dynamic_client,
+    LxdContainerClient,
+    client_name_to_class
+)
+
+
+def pytest_runtest_setup(item):
+    """Skip tests on unsupported clouds.
+
+    A test can take any number of marks to specify the platforms it can
+    run on. If a platform(s) is specified and we're not running on that
+    platform, then skip the test. If platform specific marks are not
+    specified, then we assume the test can be run anywhere.
+    """
+    all_platforms = client_name_to_class.keys()
+    supported_platforms = set(all_platforms).intersection(
+        mark.name for mark in item.iter_markers())
+    current_platform = integration_settings.PLATFORM
+    if supported_platforms and current_platform not in supported_platforms:
+        pytest.skip('Cannot run on platform {}'.format(current_platform))
+
+
+# disable_subp_usage is defined at a higher level, but we don't
+# want it applied here
+@pytest.fixture()
+def disable_subp_usage(request):
+    pass
+
+
+@pytest.fixture(scope='session', autouse=True)
+def setup_image():
+    """Setup the target environment with the correct version of cloud-init.
+
+    So we can launch instances / run tests with the correct image
+    """
+    client = dynamic_client()
+    print('Setting up environment for {}'.format(client.datasource))
```

Comment:
When it comes to tests (and pytest in particular), I actually prefer printing. I have yet to see a use-case for multiple log levels in tests. I either want everything, or I want nothing. The `-s` flag gives us a nice control for that. It's simple, doesn't require extra configuration, and doesn't require hunting down extra log files. The pytest docs actually encourage using printing this way.
https://docs.pytest.org/en/stable/capture.html#using-print-statements-for-debugging

That said, I get why people are wary of print statements. If there are strong opinions, I can switch it over.

Reply:
> I have yet to see a use-case for multiple log levels in tests.

Agreed.

> I either want everything, or I want nothing.

When it comes to integration tests, I think we always want everything. Even if a test is going to pass, I want to see that things are happening while it gets to that point.

> The -s flag gives us a nice control for that. It's simple, doesn't require extra configuration, and doesn't require hunting down extra log files.

If we configure logging to log to stdout then I think we could probably continue to use `-s`, give or take?

> The pytest docs actually encourage using printing this way. https://docs.pytest.org/en/stable/capture.html#using-print-statements-for-debugging

The example given there feels different to what we're doing. Specifically, I think that "running this module will show you precisely the output of the failing function and hide the other one" is not desirable behaviour: as mentioned above, I think we really want to emit all this information all the time, regardless of pass/fail.

> That said, I get why people are wary of print statements. If there are strong opinions, I can switch it over.

In terms of what we output, I'm not particularly fussed about it being the logging library specifically, but I do think (a) timestamped output is a killer feature for long-running tests, and (b) it is very useful to know where in the code a particular message is coming from. (For (b): if I'm debugging tests then I can just `grep` for the message, but if I'm just observing them in CI, say, then having the extra context is really helpful.) Of course, `logging` is an included battery that does these things.

Another factor to consider: our dependencies do use `logging`, so even if we don't use it, we still need to be able to configure it for debugging lower-level issues etc. My experience has been that you don't want to only do this locally (i.e. by manually configuring logging in your local tree) because the sort of CI failures that need trace-level logging to understand generally only reproduce in the CI environment.